verification/curator-service/ui/src/components/common-form-fields/Source.tsx

```diff
@@ -1,27 +1,151 @@
             <Scroll.Element name="source">
                 <fieldset>
                     <legend>Source</legend>
-                    <FastField
-                        label="Source URL"
-                        name="sourceUrl"
-                        type="text"
-                        required
-                        placeholder="https://..."
-                        component={TextField}
-                        fullWidth
-                    />
-                    <RequiredHelperText name={'sourceUrl'}></RequiredHelperText>
+                    <SourcesAutocomplete name="sourceUrl" required />
                 </fieldset>
             </Scroll.Element>
         );
     }
 }
+
+interface OriginData {
+    url: string;
+}
+
+interface SourceData {
+    _id: string;
+    origin: OriginData;
+}
+
+interface ListSourcesResponse {
+    sources: SourceData[];
+}
+
+interface SourcesAutocompleteProps {
+    name: string;
+    required?: boolean;
+}
+
+export function SourcesAutocomplete(
+    props: SourcesAutocompleteProps,
+): JSX.Element {
+    const [value, setValue] = React.useState<string | null>(null);
+    const [inputValue, setInputValue] = React.useState('');
+    const [options, setOptions] = React.useState<string[]>([]);
+    const { setFieldValue, setTouched } = useFormikContext();
+
+    const fetch = React.useMemo(
+        () =>
+            throttle(
+                async (
+                    request: { url: string },
+                    callback: (results?: SourceData[]) => void,
+                ) => {
+                    const resp = await axios.get<ListSourcesResponse>(
+                        '/api/sources',
+                        {
+                            params: request,
+                        },
+                    );
+                    callback(resp.data.sources);
+                },
+                250,
+            ),
+        [],
+    );
+
+    React.useEffect(() => {
+        let active = true;
+
+        if (inputValue.trim() === '') {
+            setOptions(value ? [value] : []);
+            return undefined;
+        }
+
+        fetch({ url: inputValue }, (results?: SourceData[]) => {
+            if (active) {
+                let newOptions = [] as string[];
+
+                if (value) {
+                    newOptions = [value];
+                }
+
+                if (results) {
+                    newOptions = [
+                        ...newOptions,
+                        ...results.map((source) => source.origin.url),
+                    ];
+                }
+
+                setOptions(newOptions);
+            }
+        });
+
+        return (): void => {
+            active = false;
+        };
+    }, [value, inputValue, fetch]);
+
+    return (
+        <Autocomplete
+            // TODO: Replace freeSolo with offering an 'Add source' dialogue.
+            freeSolo
+            itemType="SourceData"
+            options={options}
+            value={value}
```

Comment:
the other autocompletes have `defaultValue={props.initialValue}`, don't you need this to display a case source in the edit component?

Reply:
Thanks for pointing that out -- indeed, the existing value won't be displayed in the edit component. Need to use a slightly different approach, here, since this component is shared by multiple forms (with different Value interfaces). Might be best to tackle this in a separate PR.